lib/Filesystem/Domain/FileList.php

```diff
@@ -86,30 +86,50 @@ public function byExtensions(array $extensions): self
     }
 
     /**
-     * @param string[] $globPatterns
+     * @param string[] $includePatterns
+     * @param string[] $excludePatterns
      */
-    public function excludePatterns(array $globPatterns): self
+    public function includeAndExclude(array $includePatterns = [], array $excludePatterns = []): self
     {
-        return $this->filter(function (SplFileInfo $info) use ($globPatterns) {
-            foreach ($globPatterns as $pattern) {
-                if (Glob::match($info->getPathname(), $pattern)) {
-                    return false;
-                }
+        $inclusionMap = [];
+        if ($includePatterns === []) {
+            $inclusionMap['/**/*'] = true;
+        }
+
+        foreach ($includePatterns as $includePattern) {
+            $inclusionMap[$includePattern] = true;
+        }
+        foreach ($excludePatterns as $excludePattern) {
+            $inclusionMap[$excludePattern] = false;
+        }
+
+        // Sort map by keys so that more specific paths are getting matched first
+        uksort($inclusionMap, function (string $x, string $y) {
+            $partsX = explode(DIRECTORY_SEPARATOR, $x);
+            $partsY = explode(DIRECTORY_SEPARATOR, $y);
+            // Longer paths should come first
+            $countDiff = -(count($partsX) <=> count($partsY));
+            if ($countDiff !== 0) {
+                return $countDiff;
             }
 
-            return true;
+            foreach ($partsX as $i => $pathPartX) {
+                if ($pathPartX === '**' || $pathPartX === '*') {
+                    return 1;
+                }
+
+                $compare = strcmp($pathPartX, $partsY[$i]);
+                if($compare !== 0) {
+                    return $compare;
+                }
+            }
```

Comment:
can you add a commenta about what this is doing? i.e. `if the segments are the same length then...` ??

Reply:
What exactly should I comment here. If they are the same then their order doesn't matter.

Reply:
explaining what the different branches mean. It's not obvious what `==='**` and `==='*'` mean and why we return. what happens with `baz*` or `*baz` or with `*/baz/*`?

Reply:
Oh yeah wildcards within a path aren't handled. That's a valid point.